### PR TITLE
Tagged Generic Encoder/Decoder

### PR DESCRIPTION
**Pull request by mankyKitty**
The Generic functionality as it is in Waargonaut is too restrictive and has all of the same problems that the `aeson` typeclass technique has. It limits the utility of the generics and provides no additional safety, defeating the purpose of using Waargonaut.

This PR is to change the `JsonEncode` and `JsonDecode` typeclass functions to require an additional type parameter, and use this parameter to tag the function that is returned. Using the `tagged` package:

`JsonEncode t a` provides `Tagged t (Encoder f a)`
&
`JsonDecode t a` provides `Tagged t (Decoder f a)`

This allows the user to benefit from the `Generic` machinery but maintain more precise control over which instances will be used. As an example:

```haskell
{-# LANGUAGE DataKinds #-}
data EndPoints
  = InternalOnly
  | Customer

instance JsonDecode 'InternalOnly Foo where ...
instance JsonDecode 'Customer     Foo  where ...

type API = 
  "internal" :> GET [WaargJSON 'InternalOnly] Foo <|> 
  "external" :> GET [WaargJSON 'Customer] Foo
```
This is a first attempt at the API so there isn't any niceness built around it, `Tagged` has a `Wrapped` instance from lens that might be useful. Feedback appreciated!!

### CODE REVIEW

File: src/Waargonaut/Generic.hs [hunk 1]
```diff
@@ -36,11 +51,15 @@ import qualified Data.Text                     as Text
 import qualified Data.Map                      as Map
 import           Data.Scientific               (Scientific)
 
+import Data.Tagged (Tagged (..))
```
Comment (gwils) [inline src/Waargonaut/Generic.hs]: This line goes against your usual import conventions

File: src/Waargonaut/Encode.hs [hunk 2]
```diff
@@ -109,7 +109,7 @@ import qualified Data.Map                   as Map
 
 import           Data.Text                  (Text)
 
-import           Waargonaut                 (waargonautBuilder)
+import           Waargonaut.Types.Json (waargonautBuilder)
```
Comment (gwils) [inline src/Waargonaut/Encode.hs]: This line goes against your usual import conventions

File: src/Waargonaut/Generic.hs [hunk 3]
```diff
@@ -188,83 +198,96 @@ jsonInfo opts pa =
...
+    pJDec = Proxy :: Proxy (JsonDecode t)
+
     err = Left ( ConversionFailure "Generic Decoder has failed, please file a bug."
                , CursorHistory' mempty
                )
 
     failure (e,h) = modify (const h) >> throwError e
 
-    -- Pretty sure there is a better way to manage this, as my intuition about
+    -- Pretty sure there is a better way to managTaggedEncoder this, as my intuition about
```
Comment (gwils) [inline src/Waargonaut/Generic.hs]: I think this change is a mistake

Comment (mankyKitty) [reply src/Waargonaut/Generic.hs]: twas

**Pull request merged**